### PR TITLE
[Paddle] Add nn layer

### PR DESCRIPTION
**Pull request by Wong4j**
Add the following nn layers (FP16/BF16):

- FusedScaleMaskSoftmax  -  softmax.py
- DotProductAttention, MultiHeadAttention   -  attention.py
- TransformerLayer  -  transformer.py

### CODE REVIEW

**Comment by Wong4j** [comment]
Could do you please help to review? @timmoon10 
cc @jeng1220 @Tom-Zheng

**Comment by jeng1220** [comment]
LGTM.
@timmoon10 ,
All UT passed. Could you please review and merge this PR if everything looks good?

**Review by timmoon10 — APPROVED** [review]
Overall LGTM.

File: transformer_engine/paddle/layer/attention.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+
+class DotProductAttention(paddle.nn.Layer):
+    """Dot Product Attention Layer
+    BMM1 -> softmax + dropout -> BMM2
+    """
+
+    def __init__(self,
+                 norm_factor: int,
+                 attention_dropout: float = 0.0,
```
Comment (ksivaman) [inline transformer_engine/paddle/layer/attention.py:129]: Discussion; should the dropout default value be `0.0`?

Comment (ksivaman) [reply transformer_engine/paddle/layer/attention.py:129]: The `TransformerLayer` API has `0.1` default so we should stick to that.

Comment (ksivaman) [reply transformer_engine/paddle/layer/attention.py:129]: Additionally, since `DotProductAttention` and `MultiHeadAttention` are APIs, we should improve those docs (hopefully should just be copy and pasting a bunch of stuff).

**Review by ksivaman — CHANGES_REQUESTED** [review]

File: transformer_engine/paddle/layer/attention.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                 attn_mask_type: str = "causal",
+                 attention_type: str = "self",
+                 backend: str = 'transformer_engine') -> None:
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/paddle/layer/attention.py:132]: I am not sure what you are saying.

This is TE/PyTorch:
```python
class DotProductAttention(...):

    if (condition):
        # calling cuDNN fused attention
    else:
        # calling flash_attention from hazyresearch
```

This is what we made for TE/PaddlePaddle for now:
```python
class DotProductAttention(...):

    if (condition):
        # transformer_engine backed, calling cuDNN fused attention
    else:
        # paddle backend, calling PaddlePaddle Primitives 
```

And what do you mean?

As far as I know, Baidu refers to flash_attention of hazyresearch and is developing its own version, but I am not sure the current status. Once complete, it will be in [PaddlePaddle repo](https://github.com/PaddlePaddle/Paddle) instead of [hazyresearch repo](https://github.com/Dao-AILab/flash-attention).

[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/paddle/layer/attention.py:132]: The PyTorch behavior is more like:
```python
class DotProductAttention(...):
    if (condition1):
        # calling flash_attention from hazyresearch
    elif (condition2):
        # calling cuDNN fused attention
    else:
        # calling PyTorch primitives
```
This decision is made during the forward pass, not at construction time, and it can change depending on the sequence lengths. Also, it's possible for the attention to call PyTorch primitives while other Transformer layer components use TE kernels.

[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/paddle/layer/attention.py:132]: So, there are 2 things.

1. add 3rd option - `calling flash_attention op (Baidu's version)`. 
   I need to sync with Baidu to know their status. This will take some time. Also, Paddle primitives should be automatically converted to flash by the Paddle deep learning compiler if the compiler is ready, so we might not need the 3rd option. I think cuDNN is still our top priority.
2.  move the logic of kernel selecting from constructor into forward. It is easy to do because PaddlePaddle supports dynamic graph mode. Although it has maximum flexibility, it will block CUDA Graph integration if kernel can be dynamically changed at runtime, won't it?

I think both 1. and 2. need more time to have better understanding. Considering schedule pressure, I hope this PR can be merged first, then 1. and 2. can be follow up in the future.

If I misunderstand anything, please correct me.
Thanks

File: transformer_engine/paddle/layer/attention.py (hunk 1 above)
Comment (Wong4j) [reply transformer_engine/paddle/layer/attention.py]: Done. I've set the default value of `dropout` to `0.1` and added docs to `DotProductAttention` and `MultiHeadAttention`.

File: transformer_engine/paddle/layer/attention.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/paddle/layer/attention.py:157]: Yea, I think we can defer until later. Most of the backend logic in the PyTorch attention module is related to Flash Attention:
https://github.com/NVIDIA/TransformerEngine/blob/e4f9e7673b3b06d303164943f4ca33f46e865437/transformer_engine/pytorch/attention.py#L1086-L1113
It seems that this logic changes rapidly, as the cuDNN team adds kernels or fixes bugs.

File: transformer_engine/paddle/layer/attention.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+    .. note::
+
+        Argument :attr:`attention_mask` will be ignored in the `forward` call when
+        :attr:`attn_mask_type` is set to `"causal"`.
+
+    Parameters
+    ----------
+    norm_factor : float
+                    normalization factor for the attention scores.
+    attention_dropout: float, default = 0.0
```
Comment (ksivaman) [inline transformer_engine/paddle/layer/attention.py:141]:
```suggestion
    attention_dropout: float, default = 0.1
```

**Review by ksivaman — APPROVED** [review]
